### PR TITLE
[Feature]: Private S3 bucket provider

### LINKED ISSUE #5868: AWS S3 upload access denied ACL public-read
**Describe the bug**
When I try to upload a file with AWS S3 upload provider, I have a "denied access" error.

**Steps to reproduce the behavior**

1. Use npm v12.16.1
2. `npx create-strapi-app my-project --quickstart`
3. `cd my-project`
4. `yarn add strapi-provider-upload-aws-s3`
5. Create settings file at extensions/upload/config/settings.json
```json 
{
    "provider": "aws-s3",
    "providerOptions": {
      "accessKeyId": "<accesskeyid>",
      "secretAccessKey": "<secretaccesskey>",
      "region": "eu-west-3",
      "params": {
        "Bucket": "<bucket-name>"
      }
    }
  }
```
6. Go to Media Library in admin panel and upload a file

**Expected behavior**
The file should be uploaded

**Screenshots**
image

image

**System**

- Node.js version: 12.6.1
- NPM version: 6.13.4
- Strapi version: 3.0.0-beta.20
- Database: sqlite
- Operating system: Archlinux

**Additional context**
I've tested my AWS credentials through aws cli to ensure it have access to upload to the bucket.

### LINKED ISSUE #15885: Media Library: `updated_at` url param is appended incorrectly
### Required System information

- Node.js version: v16.19.0
- NPM version: 9.4.0
- Strapi version: 4.6.0
- Database: PostgreSQL
- Operating system: macOS Ventura
- Is your project Javascript or Typescript: Typescript

### Describe the bug

The url that's used to preview the file is incorrect.

### Steps to reproduce the behavior

1. Upload a file via a provider that returns a url with a `?some=param`at the end
2. The preview url just appends `?updated_at=...` at the end without regard for existing query parameters.
3. Example: `https://player.vimeo.com/progressive_redirect/playback/<some_id>/rendition/360p/file.mp4?loc=external&oauth2_token_id=...&signature=...?updated_at=2023-02-22T15:16:40.206Z`

### Expected behavior

New query parameter should be merged with existing one.

`https://player.vimeo.com/progressive_redirect/playback/<some_id>/rendition/360p/file.mp4?loc=external&oauth2_token_id=...&signature=...&updated_at=2023-02-22T15:16:40.206Z`

### Additional context

Usually this probably has no negative effect, but Vimeo will refuse to return a mp4 file if this is incorrect.

### PR DESCRIPTION
**Pull request by Marc-Roig**
### What does it do?

Describe the technical changes you did.

### Why is it needed?

Describe the issue you are solving.

### How to test it?

Provide information about the environment and the path to verify the behaviour.

### Related issue(s)/PR(s)

* resolves #15885
* resolves #5868

### CODE REVIEW

File: packages/core/upload/server/controllers/content-api.js
Comment (jhoward1994) [inline packages/core/upload/server/controllers/content-api.js:144]: Do we want to remove this part for the content api now?

Comment (Marc-Roig) [reply packages/core/upload/server/controllers/content-api.js:144]: yes!

**Comment by Marc-Roig** [comment]
I think there are two things left to do:
- Fix front tests
- Do not sign images if they are uploaded using another bucket. 
	- Is this possible? Maybe before signing checking if the image has a different url?

File: packages/core/upload/server/services/extensions/content-manager/entity-manager.js
Comment (nathan-pichon) [inline packages/core/upload/server/services/extensions/content-manager/entity-manager.js:69]: I think you still need to call the entityManage mapEntity method.

File: packages/providers/upload-aws-s3/lib/index.js
Comment (nathan-pichon) [inline packages/providers/upload-aws-s3/lib/index.js:9]:
```suggestion
const { get } = require('lodash/fp');
```

File: packages/core/content-manager/server/services/entity-manager.js
Comment (nathan-pichon) [inline packages/core/content-manager/server/services/entity-manager.js:99]: Could you check that the type of entities is single type?

Comment (Marc-Roig) [reply packages/core/content-manager/server/services/entity-manager.js:99]: That make sense,
Just a point, that would require to load the whole uid model in this function. 
We thought it was a bit of an overhead , but it makes sense to be more explicit here. wdyt @jhoward1994 ?

File: packages/core/upload/server/services/extensions/content-manager/entity-manager.js
Comment (nathan-pichon) [inline packages/core/upload/server/services/extensions/content-manager/entity-manager.js:25]: Should we return values instead of using a function that set the value ? 🤔 Isn't this hard to maintain / read / understand to not return anything and call sort of a callback?

File: packages/core/upload/server/services/file.js
Comment (nathan-pichon) [inline packages/core/upload/server/services/file.js:47]: Is this doing mutation? 😅

File: packages/core/upload/server/services/extensions/content-manager/entity-manager.js
Comment (Marc-Roig) [reply packages/core/upload/server/services/extensions/content-manager/entity-manager.js:25]: I personally find this much easier to follow, as you can set but also remove fields (or more actions)

This is how the traverse entity util is defined in strapi and we use it for sanitizing and a couple more places. We are starting discussions to refactor it, so we could see how we can improve it together!

File: packages/core/upload/server/services/file.js
Comment (Marc-Roig) [reply packages/core/upload/server/services/file.js:47]: a mutation on the cloned entity, do you think it's cleaner to not do it this way? 

I was trying to find other alternatives but I found this one easy to read. wdyt?

**Review by nathan-pichon — APPROVED** [review]
Some comments, but nothing that block this PR. 👍

File: packages/core/upload/server/services/file.js
Comment (nathan-pichon) [reply packages/core/upload/server/services/file.js:47]: Mutation are to avoid, we should never have some. It's really complicated to follow where things are being modified with mutation 🙁

Comment (Marc-Roig) [reply packages/core/upload/server/services/file.js:47]: but don't you think in this case it's clean to follow?

I feel it's easy to follow when things are being modified if you are mutating an object you defined in the same function.
Otherwise  Iwould need a more complicated map iterating all fields.

And because of tyring to avoid mutations, would make the code harder to read.

File: packages/providers/upload-aws-s3/lib/index.js
Comment (Marc-Roig) [reply packages/providers/upload-aws-s3/lib/index.js:94]: didn't know that 👀 I will take a look

File: packages/core/upload/server/services/file.js
Comment (jhoward1994) [reply packages/core/upload/server/services/file.js:47]: to me mutating `signedFile` is okay in this case. It is a clone of the file entity and is defined within the scope of this function so shouldn't have an impact elsewhere

Comment (nathan-pichon) [reply packages/core/upload/server/services/file.js:47]: I understand what you say, and I would agree for the part where the mutation is contained in the function's body (and the function is relatively small and easy to read).
Le problème c'est que je suis tombé de nombreuses fois sur ce type de mutation où au final, la fonction a évolué avec le temps et tout est devenu illisible et incompréhensible, engendrant des bugs à se casser la tête pour trouver la source à cause d'une mutation.
La fonction `signUrl(file)` mute son argument `file`.
Pourquoi ne pas retourner l'url signée ?
```javascript
const signUrl = async (file) => {
  const signedUrl = await provider.getSignedUrl(file);
  return signedUrl.url;
};
// Sign each file format
signedFile.url = await signUrl(signedFile);
if (file.formats) {
  signedFile.formats = await mapAsync(Object.values(signedFile.formats), async (format) => {
    const url = await signUrl(format);
    return { ...format, url };
  });
}
```

File: packages/core/content-manager/server/services/entity-manager.js
Comment (jhoward1994) [reply packages/core/content-manager/server/services/entity-manager.js:99]: I think safety for single types is already covered in the entity service. see:

https://github.com/strapi/strapi/blob/19e9f952efb08c9dca8f261132479a239d7659d8/packages/core/strapi/lib/services/entity-service/index.js#L85

However we could change this code block to 
```
    if (entities?.results || Array.isArray(entities)) {
      const mappedResults = await mapAsync(entities?.results || entities, (entity) =>
        this.mapEntity(entity, uid)
      );
      return { ...entities, results: mappedResults };
    }
    // if entity is single type
    return this.mapEntity(entities, uid);
```

wdyt?

File: docs/docs/core/upload/backend/providers/intro.md
Comment (jihyukbae) [inline docs/docs/core/upload/backend/providers/intro.md:25]: Hi, driving by with a quick question. According to this, it seems like this will be compatible with other providers (e.g. [this one](https://github.com/strapi-community/strapi-provider-upload-google-cloud-storage) from the community for Google Cloud Storage). 

Once this PR is merged, would enabling access to private storage be as simple as adding these two new methods to the existing provider?

Comment (Marc-Roig) [reply docs/docs/core/upload/backend/providers/intro.md:25]: Hello :) Yes! this PR will make possible adapting other providers (like GCS) by using the `isPrivate()` and `getSignedUrl(file)` methods.

We will create an experimental release this next week. If anyone is interested on adapting one of the GCS providers and test if this works it would be great to test this feature too!

Comment (jihyukbae) [reply docs/docs/core/upload/backend/providers/intro.md:25]: Awesome, thank you! I'm definitely interested in testing out the experimental release :D

**Review by innerdvations — COMMENTED** [review]
The "copy link" in the content manager for a related media item in a private bucket includes the signed url parameters. I don't know if that's intended, but since I understood it as those parameters only being for the viewing of the image within the admin ui I wasn't sure if that link should only be the raw image url.

File: packages/providers/upload-aws-s3/lib/index.js
Comment (Marc-Roig) [reply packages/providers/upload-aws-s3/lib/index.js:25]: Thank you for the QA @innerdvations !

We changed the lodash get to use `lodash/fp` and it seems it does not have the 'default' argument. I have created a pr for it!

> The "copy link" in the content manager for a related media item in a private bucket includes the signed url parameters. I don't know if that's intended, but since I understood it as those parameters only being for the viewing of the image within the admin ui I wasn't sure if that link should only be the raw image url.

Good catch! We originally did not think bout this, but I think we could leave it this way, as it will provide users a way to share the image, which will expire in a certain amount of time.

Comment (innerdvations) [reply packages/providers/upload-aws-s3/lib/index.js:25]: Sounds good. I wonder if we should document that the links will only last that long? I could see content editors attempting to use that link somewhere in their content.

Comment (Marc-Roig) [reply packages/providers/upload-aws-s3/lib/index.js]: That's a good point indeed, this is a use case that we will debate in the team on further iterations.

**Comment by Marc-Roig** [comment]
Everything seems to be ready :)  I have created this experimental release, for any one interested in testing the feature.
`0.0.0-15b41fe7a432ecf2ed59c752e55b6f867f3ade54`

ping @jihyukbae .


<details>
<summary>FAQ - How to use the private s3 bucket</summary>

## **- How can I configure the s3 provider as private?**
If your bucket is configured to be private, you can set the `ACL` option to `private` in the `params` object. This will ensure that the signed URL is generated with the correct permissions.

You can also define the expiration time of the signed URL by setting the `signedUrlExpires` option in the `params` object. The default value is 7 days.

To set up the private S3 bucket, you need to set in config/plugins.js:
```js
  upload: {
    config: {
      provider: 'aws-s3',
      providerOptions: {
        accessKeyId: env('AWS_ACCESS_KEY_ID'),
        secretAccessKey: env('AWS_ACCESS_SECRET'),
        region: env('AWS_REGION'),
        params: {
          ACL: 'private',
          Bucket: env('AWS_BUCKET'),
        },
      },
    },
  },
```

Strapi has a default security middleware that has a very strict contentSecurityPolicy that limits loading images and media to "'self'" only, see the example configuration on the [provider page](https://www.npmjs.com/package/@strapi/provider-upload-aws-s3) or the [middleware documentation](https://docs.strapi.io/dev-docs/configurations/middlewares#security) for more information.

You have to set up, in config/middlewares.js:

```js
module.exports = [
  ...
  {
    name: 'strapi::security',
    config: {
      contentSecurityPolicy: {
        useDefaults: true,
        directives: {
          'connect-src': ["'self'", 'https:'],
          'img-src': [
            "'self'",
            'data:',
            'blob:',
            'bucket-name.s3.region.amazonaws.com'// <--- set your bucket url
          ],
          'media-src': [
            "'self'",
            'data:',
            'blob:',
            'bucket-name.s3.region.amazonaws.com'// <--- set your bucket url
          ],
          upgradeInsecureRequests: null,
        },
      },
    },
  },
...
];
```

## **Can I share links that appear on the Content Manager?**
    
    Yes, you can share them and the links will expire based on the `signedUrlExpires` configuration mentioned in the previous question.
    
## **Does the Content-API have authenticated media?**
    
    Meaning I can open images (or any other file) that come from the content-api.
    
    No, it does not authenticate the media. The user has to use their own authentication method for it.
    
## **If I switch my current s3 bucket to be private, will it still work for the images that I previously uploaded when it was public?**
    
    Yes
    
## **Do we plan to support other such providers in the future (Like Google Cloud) for example?**
    
    Yes, we plan to support other providers in the future, **but development work will be necessary as serving private files works differently for different providers**. The overall approach will remain the same, with config for private providers being added in /config/plugins.js. We have prioritized supporting Amazon S3 buckets for now, as that has been the most popular request.

</details>

File: docs/docs/core/upload/backend/providers/intro.md
Comment (Marc-Roig) [reply docs/docs/core/upload/backend/providers/intro.md:25]: have you added the domain of your bucket in the middlewares?

example:
	```
	./config/middlewares.js

module.exports = [
  // ...
  {
    name: 'strapi::security',
    config: {
      contentSecurityPolicy: {
        useDefaults: true,
        directives: {
          'connect-src': ["'self'", 'https:'],
          'img-src': [
            "'self'",
            'data:',
            'blob:',
            'dl.airtable.com',
            'yourBucketName.s3.yourRegion.amazonaws.com',
          ],
          'media-src': [
            "'self'",
            'data:',
            'blob:',
            'dl.airtable.com',
            'yourBucketName.s3.yourRegion.amazonaws.com',
          ],
          upgradeInsecureRequests: null,
        },
      },
    },
  },
  // ...
]
```

Comment (Marc-Roig) [reply docs/docs/core/upload/backend/providers/intro.md:25]: and thank you for actually testing this on the release branch , this is very helpfull.

Comment (jihyukbae) [reply docs/docs/core/upload/backend/providers/intro.md:25]: This is what I currently have - would including `storage.googleapis.com` not be sufficient?
Screenshot 2023-03-15 at 12 15 33 PM

Comment (Marc-Roig) [reply docs/docs/core/upload/backend/providers/intro.md:25]: can you share the `host` part of one of the urls that the api returns? or just the url if possible

Comment (Marc-Roig) [reply docs/docs/core/upload/backend/providers/intro.md:25]: for example for a s3 bucket I have

`<bucket-name>.s3.eu-west-3.amazonaws.com`

Comment (jihyukbae) [reply docs/docs/core/upload/backend/providers/intro.md:25]: Sure, the url looks like: https://storage.googleapis.com/[bucket name]/thumbnail_cat2.jpg?X-Goog-Algorithm=...(signature info)

Comment (Marc-Roig) [reply docs/docs/core/upload/backend/providers/intro.md:25]: then storage.googleapis.com should be enough for your case. 

So, if you open that image directly from your browser it works. We could say it is being signed correctly.

It is on the frontend side that it's not even making the request to load the image right? It really sounds a problem from the middlewares config, but we discarded that.

Does the GCP work for you with a public configuration?

Comment (Marc-Roig) [reply docs/docs/core/upload/backend/providers/intro.md:25]: ah, once you updated the strapi version to use the nightly version, have you built your project?

Comment (jihyukbae) [reply docs/docs/core/upload/backend/providers/intro.md:25]: re: public configuration: i haven't tried it myself, but assumed it would from reading others' experiences.

Hmm, I'm relatively certain I've rebuilt the project, but I can try again as a sanity check.

Comment (jihyukbae) [reply docs/docs/core/upload/backend/providers/intro.md:25]: Rebuilt the project and seeing the same behavior.

re: public config: Sure, I can give that a try. It still seems odd to me that no request is being sent to fetch the image though...

Comment (Marc-Roig) [reply docs/docs/core/upload/backend/providers/intro.md:25]: yeah , that only should be the case when it's blocked by the strapi::security middleware

Comment (Marc-Roig) [reply docs/docs/core/upload/backend/providers/intro.md:25]: is there another error on the browser console?

Comment (jihyukbae) [reply docs/docs/core/upload/backend/providers/intro.md:25]: hm... no, not seeing any errors there.

Comment (Marc-Roig) [reply docs/docs/core/upload/backend/providers/intro.md:25]: I would first see if it happens for a public config then, it will point us if this is a config problem or there is smething wrong with this branch.

Comment (jihyukbae) [reply docs/docs/core/upload/backend/providers/intro.md:25]: sounds good, will circle back here once i confirm what happens with a public config. thanks!

Comment (Marc-Roig) [reply docs/docs/core/upload/backend/providers/intro.md:25]: thank you for testing this !

Comment (jihyukbae) [reply docs/docs/core/upload/backend/providers/intro.md:25]: When I have `isPrivate()` in the provider return false, I can see that requests are made to fetch the image that fail with a 403 error (as expected). Since as you mentioned, we've ruled out the security middleware, this leads me to think that something may be wrong in this branch. What do you think?

And sure! Thank you for working on this :)

Comment (Marc-Roig) [reply docs/docs/core/upload/backend/providers/intro.md:25]: I will try to replicate your setup this week and see what it is 🙂

Comment (jihyukbae) [reply docs/docs/core/upload/backend/providers/intro.md:25]: Thanks so much! Would be happy to clarify anything further and/or jump on a call if you'd like.

Comment (Marc-Roig) [reply docs/docs/core/upload/backend/providers/intro.md:25]: Hello :) I have managed to test this actually, your code was actually super useful!

Everything seems to work ok on my setup :/  I am able to authenticate images and display them in the Media Library.

I find weird that only when the image is authenticated your browser doesn't make the request, could it be that something else is blocking it?

We could jump in a call next week monday if you are available :) 
Also, I would use this opportunity to open a pr in https://github.com/strapi-community/strapi-provider-upload-google-cloud-storage to add this functionality.

Comment (jihyukbae) [reply docs/docs/core/upload/backend/providers/intro.md:25]: Interesting...it does seem weird that on my end, the browser doesn't make the request only when the image needs auth.

Sure, I'd be more than happy to jump on a call next Monday! Would sometime between 9am-1:30pm ET work?

And yes, also happy to open a PR in the GCS provider repo once I confirm the functionality on my end.

Comment (jihyukbae) [reply docs/docs/core/upload/backend/providers/intro.md:25]: Actually, Monday may not be a good time - maybe we can discuss further on Discord? Is your username there Marc-André?

Comment (Marc-Roig) [reply docs/docs/core/upload/backend/providers/intro.md:25]: @jihyukbae  my discord user: MarcRoig#5524 

feel free to ping me whenever you are available :)

Comment (Marc-Roig) [reply docs/docs/core/upload/backend/providers/intro.md:25]: @jihyukbae I created the PR in the meantime https://github.com/strapi-community/strapi-provider-upload-google-cloud-storage/pull/165


Feel free to add any comment in there!! Let's see if we can find more people to test this

**Pull request merged**
